scRNA-seq/02-normalizing_scRNA-seq.Rmd

```diff
@@ -0,0 +1,255 @@
+`SingleCellExperiment` object.
+
+For more information on `SingleCellExperiment` objects, check out this [excellent vignette](https://bioconductor.org/packages/devel/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html)
+and/or [Amezquita et al. _bioRxiv._ 2019.](https://www.biorxiv.org/content/10.1101/590562v1)
+
+## Normalization of count data 
+
+In whatever data we are working with, we are always looking to maximize biological 
+variance and minimize technical variance. 
+This is where normalization methods usually come into the workflow.
+With this high amount of zeroes in our data, this makes normalization particularly
+tricky.
+So in order to account for these zeroes, we normalize cells in groups with other 
+cells like them; a method introduced in [Lun et al 2016](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0947-7).
+
+*These steps may take a minute or so.*
+
+```{r}
+# Step 1) Group cells with other like cells by clustering.  
+qclust <- scran::quickCluster(sce_raw)
+
+# Step 2) Compute sum factors for each cell cluster grouping.  
+sce_raw <- scran::computeSumFactors(sce_raw, clusters = qclust)
+
+### TODO: This is now deprecated- update to latest recommendations ###
+
+# Step 3) Normalize using these pooled sum factors and log transform. 
+sce_norm <- scater::normalize(sce_raw)
+```
+
+## Compare normalized data to count data
+
+One way to determine whether our normalization yields biologically relevant 
+results, is to plot it and see if similarly labeled samples and cells end up 
+together.
+Because plotting 1000's of genes together isn't practical, we will reduce the 
+dimensions of our data using Principal Components Analysis. 
+Note that the `t()` function is being used here to transpose the `matrix`, so 
+it is samples x genes. 
+We could however, obtain PCA scores for each gene if we didn't transpose our
+data.frame first. 
+
+```{r}
+# Use PCA for dimension reduction of cells' raw data
+count_pca <- prcomp(t(counts(sce_raw)))
+
+# Use PCA for dimension reduction of cells' scran normalized data
+norm_pca <- prcomp(t(logcounts(sce_norm)))
+```
+
+Now we will retrieve the PCA scores, by using `$x` at the end of our `prcomp`
+objects.
+
+```{r Graph PCA}
+# Make the counts PC scores into a dataframe with cell type labels for easy graphing 
+count_pca_scores <- data.frame(count_pca$x[, 1:2],
+                               geo_accession = rownames(count_pca$x)) %>%
+  dplyr::left_join(dplyr::select(sc_metadata, 
+                                 geo_accession, 
+                                 cell_type = cell.type.ch1))
+
+# Set up the normalized PC scores in the same way
+norm_pca_scores <- data.frame(norm_pca$x[, 1:2],
+                              geo_accession = rownames(norm_pca$x)) %>%
+  dplyr::left_join(dplyr::select(sc_metadata, 
+                                 geo_accession, 
+                                 cell_type = cell.type.ch1))
+```
```

Comment:
I do like the idea of covering `prcomp()` (over using `runPCA()` from `scater`), but I am less attached to the idea of this plot _overall_.  I also like the we-plot-now-you-plot (™️ ?) format. We should consider what we're hoping to illustrate in this section.

Reply:
At this stage, I was keeping much of the content for the second notebook the same; but with the updated data. Expect more changes in a separate PR. I think this pair of plots does nicely illustrate *part* of what normalization does, but I would like to hear thoughts about about what the "best" plots to illustrate the effect of normalization might be.

Reply:
> At this stage, I was keeping much of the content for the second notebook the same; but with the updated data. Expect more changes in a separate PR.

Yep, understood.

My concern with this is that PCA plots are perhaps conflating multiple changes (e.g., the normalization will account for cell-specific factors but also put things on a log scale) and we're not scaling in the way we run PCA currently / I expect there may be outliers in the counts.